omnigan/trainer.py

```diff
@@ -692,28 +728,115 @@ def get_painter_loss(self, multi_domain_batch):
             )
             step_loss += update_loss
 
-            update_loss = self.losses["G"]["p"]["tv"](fake_flooded)
+            update_loss = self.losses["G"]["p"]["tv"](fake_flooded * m)
             self.logger.losses.generator.p.tv = update_loss.item()
             step_loss += update_loss
 
             update_loss = (
-                self.losses["G"]["p"]["context"](fake_flooded * (1.0 - m), masked_x)
+                self.losses["G"]["p"]["context"](fake_flooded, x, m)
                 * lambdas.G["p"]["context"]
             )
 
             self.logger.losses.generator.p.context = update_loss.item()
             step_loss += update_loss
 
+            # GAN Losses
             fake_d_global = self.D["p"]["global"](fake_flooded)
             fake_d_local = self.D["p"]["local"](fake_flooded * m)
-            update_loss = (
-                self.losses["G"]["p"]["gan"](fake_d_global, True)
-                + self.losses["G"]["p"]["gan"](fake_d_local, True)
-            ) * lambdas.G["p"]["gan"]
 
-            self.logger.losses.generator.p.gan = update_loss.item()
+            real_d_global = self.D["p"]["global"](x)
+
+            # Note: discriminator returns [out_1,...,out_num_D] outputs
+            # Each out_i is a list [feat1, feat2, ..., pred_i]
+
+            self.logger.losses.generator.p.gan = 0
+
+            num_D = len(fake_d_global)
+            for i in range(num_D):
+                # Take last element for GAN loss on discrim prediction
+                update_loss = (
+                    (
+                        self.losses["G"]["p"]["gan"](fake_d_global[i][-1], True)
+                        + self.losses["G"]["p"]["gan"](fake_d_local[i][-1], True)
+                    )
+                    * lambdas.G["p"]["gan"]
+                    / num_D
+                )
+
+                self.logger.losses.generator.p.gan += update_loss.item()
+
             step_loss += update_loss
 
+            # Feature matching loss (only on global discriminator)
+            # Order must be real, fake
+            if self.opts.dis.p.get_intermediate_features:
+                update_loss = (
+                    self.losses["G"]["p"]["featmatch"](real_d_global, fake_d_global)
+                    * lambdas.G["p"]["featmatch"]
+                )
+                if isinstance(update_loss, float):
```

Comment:
how could it be a float? why would `self.losses["G"]["p"]["featmatch"](real_d_global, fake_d_global)` return a float?

Reply:
It's a float by default. So if nothing happens (for example if we don't return the features from the discriminator at all) then we get a float